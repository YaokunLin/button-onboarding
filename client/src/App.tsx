import React from 'react';
import logo from './logo.svg';
import './App.css';
import graphql from 'babel-plugin-relay/macro';
import { useLazyLoadQuery } from 'react-relay';

const AppQuery = graphql`
query AppQuery {
  allTodos {
    edges {
      node {
        completed
        dateCreated
        dateUpdated
        userUid
      }
    }
  }
}
`;

function App() {
  const data = useLazyLoadQuery(
    AppQuery,
    {},
  );
  console.log(data)
  return (
    <div className="App">
      hello world
    </div>
  );
}

export default App;
