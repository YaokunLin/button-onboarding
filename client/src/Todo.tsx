import './App.css';
import graphql from 'babel-plugin-relay/macro';
import { ConnectionHandler, useFragment, useMutation } from 'react-relay';
import { TodoFragment$key } from './__generated__/TodoFragment.graphql';
import { RecordSourceSelectorProxy } from 'relay-runtime/lib/store/RelayStoreTypes';
import { useState } from 'react';
import { DeleteTodoMutation } from './mutations/DeleteTodoMutation';

const TodoFragment = graphql`
  fragment TodoFragment on Todo {
    __id
    taskUid
    completed
    dateCreated
    dateUpdated
    userUid
    task
  }
`;

const TodoMutation = graphql`
  mutation TodoMutation(
    $taskUid: UUID!,
    $completed: Boolean!,
    $task: String!,

  ) @raw_response_type {
    updateTodoByTaskUid(
      input: {todoPatch: {completed: $completed, task:$task}, taskUid: $taskUid}
    ) {
      todo {
        ...TodoFragment
      }
    }
  }
`;



type Props = {
  todoProp: TodoFragment$key;
  };


function Todo({todoProp} : Props) {
  const fetchedTodo = useFragment(
    TodoFragment,
    todoProp,
  );
  const [commitMutation, isMutationInFlight] = useMutation(TodoMutation);
  const [taskTxt, setTaskTxt] = useState(fetchedTodo.task);
  const [isCompleted, setIsCompleted] = useState(fetchedTodo.completed);
  const [deleteMutation, isDeleteMutationInFlight] = useMutation(DeleteTodoMutation);
  const handleOnChange = () => commitMutation({
    variables: {
      taskUid: fetchedTodo.taskUid,
      completed: isCompleted,
      task: taskTxt,
    },
    updater: (store: RecordSourceSelectorProxy, response: any) => {
      console.log("updated")
      const payload = store.getRootField('updateTodoByTaskUid');
      if (payload && fetchedTodo) {
        const updatedTodo = payload.getLinkedRecord('todo');
   
        if (updatedTodo && fetchedTodo.__id) {
          const cachedTodo = store.get(fetchedTodo.__id);
       
          if (cachedTodo) {
            cachedTodo.copyFieldsFrom(updatedTodo);
          }
        }
      }
    },
  })


 

  const handleDelete = () => {
    deleteMutation({
      variables: {
        taskUid: fetchedTodo.taskUid,
      },
      updater: (store: RecordSourceSelectorProxy) => {
        const userRecord = store.getRoot();
        const todosConnection = ConnectionHandler.getConnection(userRecord, 'TodosQuery_allTodos');
        console.log(fetchedTodo.__id)

        if (todosConnection) {
          ConnectionHandler.deleteNode(todosConnection, fetchedTodo.__id);
        }
      },
    });
  };


  return (
    <tr style={{background:fetchedTodo.completed? "green": "red"}}>
       <th>
          {taskTxt}    
      </th>
      <th>
        <input 
              type="checkbox" 
              name={fetchedTodo.task} 
              checked={isCompleted ? true : false} 
              onChange={(e) => setIsCompleted(v=>!v)}
              disabled={isMutationInFlight}
        />
     
          <input 
            type="text" 
            value={taskTxt}
            onChange={(evt)=>setTaskTxt(evt.target.value)}
            disabled={isMutationInFlight}
          />

          <button 
            onClick={handleOnChange}
            disabled={isMutationInFlight}
            >
              Update
          </button>
      </th>

      

      <th>
          <button 
            onClick={handleDelete}
            disabled={isDeleteMutationInFlight}
            >
              X
          </button>
      </th>
      
      
      
    </tr>
  );
}

export default Todo;
