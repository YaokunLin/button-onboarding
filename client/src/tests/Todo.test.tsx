import React, { Suspense } from 'react';
import { render, screen } from '@testing-library/react';
import App from '../App';
import {createMockEnvironment, MockPayloadGenerator} from 'relay-test-utils';
import ReactTestRenderer from 'react-test-renderer'; 
import { RelayEnvironmentProvider, useLazyLoadQuery } from 'react-relay';
import graphql from 'babel-plugin-relay/macro';
import type {TodosTestQuery as TodosTestQueryType} from './__generated__/TodosTestQuery.graphql';
import Todo from '../Todo';


test('renders Loading', async () => {
    const environment = createMockEnvironment();
  
    render(
      <RelayEnvironmentProvider environment={environment}>
        <Suspense fallback={"Loading..."}>
          <App />
        </Suspense>
      </RelayEnvironmentProvider>,
    );
  
    const textElement = await screen.findByText("Loading...");
    expect(textElement).toBeInTheDocument();
  });

  test('Fragment', () => {
    const environment = createMockEnvironment();
    const TestRenderer = () => {  
        const TodosTestQuery = graphql`
        query TodosTestQuery($count: Int)  
        {
          allTodos(first: $count) @connection(key: "TodosQuery_allTodos") {
            __id
            edges {
              node {
                ...TodoFragment
                taskUid
              }
            }
          }
        }
        `;
    const data = useLazyLoadQuery<TodosTestQueryType>(
            TodosTestQuery,
            {count: 100},
          );
    const allTodos = data.allTodos;
    const todos = allTodos?.edges;
    if (todos && todos[0] && todos[0].node){
        return <Todo todoProp={todos[0].node} />
    }
  
    return <>no todos</>
    };
  
    const renderer = ReactTestRenderer.create(
      <RelayEnvironmentProvider environment={environment}>
        <Suspense fallback="Loading...">
          <TestRenderer />
        </Suspense>
      </RelayEnvironmentProvider>
    );
  
    // Wrapping in ReactTestRenderer.act will ensure that components
    // are fully updated to their final state.
    ReactTestRenderer.act(() => {
      environment.mock.resolveMostRecentOperation(operation =>
        MockPayloadGenerator.generate(operation),
      );
    });
  
    expect(renderer).toMatchSnapshot();
  });