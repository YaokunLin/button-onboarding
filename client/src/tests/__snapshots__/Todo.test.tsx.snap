// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fragment 1`] = `"Loading..."`;
